server/auth/azuread.js

```diff
@@ -0,0 +1,211 @@
+var tenant = process.env.AZUREAD_TENANT_ID;
+var authorityUrl = authorityHostUrl + tenant;
+var redirectUri = `${process.env.URL}/auth/azuread.callback`;
+var templateAuthzUrl = `${
+  authorityUrl
+}//oauth2/authorize?response_type=code&client_id=${clientId}&&redirect_uri=${
+  redirectUri
+}&state=<state>`;
+
+var createAuthorizationUrl = state => {
+  return templateAuthzUrl.replace('<state>', state);
+};
+
+var aqcuireTokenAsync = (code, redirectUri, clientId, clientSecret) => {
+  return new Promise((resolve, reject) => {
+    var authenticationContext = new AuthenticationContext(authorityUrl);
+    authenticationContext.acquireTokenWithAuthorizationCode(
+      code,
+      redirectUri,
+      'https://graph.microsoft.com',
+      clientId,
+      clientSecret,
+      (err, response) => {
+        if (err) {
+          reject(err);
+        }
+
+        resolve(response);
+      }
+    );
+  });
+};
+
+// start the oauth process and redirect user to Microsoft AzureAD
+router.get('azuread', async ctx => {
+  const buffer = await crypto.randomBytes(48);
+  const token = buffer
+    .toString('base64')
+    .replace(/\//g, '_')
+    .replace(/\+/g, '-');
+  ctx.cookies.set('authstate', token);
+  var authorizationUrl = createAuthorizationUrl(token);
+  ctx.redirect(authorizationUrl);
+});
+
+// signin callback from Microsoft AzureAD
+router.get('azuread.callback', auth({ required: false }), async ctx => {
+  const { code, state } = ctx.request.query;
+  ctx.assertPresent(code, 'code is required');
+
+  if (ctx.cookies.get('authstate') !== state) {
+    ctx.redirect('/?notice=auth-error');
+    return;
+  }
+
+  var response = await aqcuireTokenAsync(
+    code,
+    redirectUri,
+    clientId,
+    clientSecret
+  );
+
+  if (!response) {
+    ctx.redirect('/?notice=auth-error');
+    return;
+  }
+
+  const accessToken = response.accessToken;
+  const clientOptions = {
+    authProvider: new AzureADProvider(accessToken),
+  };
+
+  const client = Client.initWithMiddleware(clientOptions);
+
+  let userId = '';
+  let azureId = '';
+  let hashedOrganizationId = '';
+  let teamName = '';
+  let userName = '';
+  let userEmail = '';
+  let userAvatar = '';
+
+  try {
+    try {
+      const orgResource = await client.api('/organization').get();
+
+      if (!orgResource) {
+        ctx.redirect('/?notice=azuread-no-organization');
+        return;
+      }
+
+      const hash = crypto.createHash('sha256');
+      hash.update(orgResource.value[0].id);
+      azureId = orgResource.value[0].id;
+      hashedOrganizationId = hash.digest('hex');
+      teamName = orgResource.value[0].displayName;
+    } catch (error) {
+      throw error;
+    }
+
+    const userResource = await client.api('/me').get();
+
+    if (!userResource) {
+      ctx.redirect('/?notice=azuread-no-userinfo');
+      return;
+    }
+
+    userId = userResource.id;
+    userName = userResource.displayName;
+    userEmail = userResource.mail;
```

Comment:
```suggestion
    userEmail = userResource.mail || userResource.userPrincipalName;
```

`userResource.mail` could be empty if the AAD user doesn't have an email inbox. That will cause an error in `findOrCreate` because it's a required field. 

https://docs.microsoft.com/en-us/azure/active-directory/hybrid/plan-connect-userprincipalname
https://stackoverflow.com/a/47480542

Reply:
Thank you, I have a mailbox, but I don't use it, haven't checked the docs TBH.